File: tools/c/overlay_extract.c
// Complile using - gcc overlay.c -o overlay.exe -lversion (or just gcc overlay.c -o overlay.exe)
// Author: Ramadhan Adam Zome, @27.7.2026

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <windows.h>

int main(int argc, char *argv[]) {
    if (argc < 3) {
        printf("Usage: overlay.exe <input.exe> <output.bin>\n");
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) {
        printf("[-] Cannot open file\n");
        return 1;
    }

    // Read DOS header
    IMAGE_DOS_HEADER dos;
    fread(&dos, sizeof(dos), 1, f);

    // Check MZ signature
    if (dos.e_magic != IMAGE_DOS_SIGNATURE) {
        printf("[-] Not a valid DOS/PE file\n");
        fclose(f);
        return 1;
    }

    // Jump to PE header using e_lfanew
    fseek(f, dos.e_lfanew, SEEK_SET);
    IMAGE_NT_HEADERS nt;
    fread(&nt, sizeof(nt), 1, f);

    // Check PE signature
    if (nt.Signature != IMAGE_NT_SIGNATURE) {
        printf("[-] Not a valid PE file\n");
        fclose(f);
        return 1;
    }

    // This reads IMAGE_NT_HEADERS64 (PE32+) since we compile for x64.
    // A 32-bit PE has a different OptionalHeader layout/size, so check Magic.
    if (nt.OptionalHeader.Magic != IMAGE_NT_OPTIONAL_HDR64_MAGIC) {
        printf("[-] Only 64-bit PE files are supported\n");
        fclose(f);
        return 1;
    }

    // Find last section end
    uint32_t pe_end = nt.OptionalHeader.SizeOfHeaders;
    for (int i = 0; i < nt.FileHeader.NumberOfSections; i++) {
        IMAGE_SECTION_HEADER sec;
        fread(&sec, sizeof(sec), 1, f);
        uint32_t sec_end = sec.PointerToRawData + sec.SizeOfRawData;
        if (sec_end > pe_end) pe_end = sec_end;
    }

    // Digital signature also sits after the last section on disk,
    // so it looks like overlay if we don't account for it
    IMAGE_DATA_DIRECTORY sec_dir = nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_SECURITY];
    if (sec_dir.VirtualAddress != 0) {
        uint32_t sig_end = sec_dir.VirtualAddress + sec_dir.Size;
        printf("[!] Digital signature found at byte %u, size %u\n", sec_dir.VirtualAddress, sec_dir.Size);
        if (sig_end > pe_end) pe_end = sig_end;
    }

    // Get file size
    fseek(f, 0, SEEK_END);
    uint32_t file_size = ftell(f);

    if (file_size <= pe_end) {
        printf("[-] No overlay\n");
        fclose(f);
        return 1;
    }

    // Extract overlay
    uint32_t overlay_size = file_size - pe_end;
    printf("[+] Overlay size: %u bytes\n", overlay_size);

    fseek(f, pe_end, SEEK_SET);
    uint8_t *overlay = malloc(overlay_size);
    if (!overlay) {
        printf("[-] malloc failed\n");
        fclose(f);
        return 1;
    }
    if (fread(overlay, overlay_size, 1, f) != 1) {
        printf("[-] Failed to read overlay data\n");
        free(overlay);
        fclose(f);
        return 1;
    }
    fclose(f);

    // Write overlay
    FILE *out = fopen(argv[2], "wb");
    if (!out) {
        printf("[-] Cannot open output file\n");
        free(overlay);
        return 1;
    }
    fwrite(overlay, overlay_size, 1, out);
    fclose(out);

    printf("[+] Saved to %s\n", argv[2]);
    free(overlay);
    return 0;
}

File: tools/py/overlay_extract.py
# tools/overlay_extract.py
# Extracts any data appended after the last PE section (the "overlay"),
# commonly used by malware to hide payloads. Skips the Authenticode
# signature block if present, since it also lives past the last section.

import sys
import pefile

def extract_overlay(exe_path, output_path):
    print(f"[*] Opening: {exe_path}")

    try:
        pe = pefile.PE(exe_path, fast_load=True)

        # Headers end here (DOS header + PE headers + optional header)
        headers_end = pe.OPTIONAL_HEADER.SizeOfHeaders

        # Find where last section ends on disk
        sections_end = 0
        for section in pe.sections:
            section_end = section.PointerToRawData + section.SizeOfRawData
            if section_end > sections_end:
                sections_end = section_end

        # PE ends at whichever is larger (guards against packed/odd PEs
        # with missing or zero sections)
        pe_end_offset = max(headers_end, sections_end)
        print(f"[+] PE ends at byte: {pe_end_offset}")

        # Authenticode signature also sits after the last section on disk,
        # so it looks like overlay if we don't account for it
        sec_dir = pe.OPTIONAL_HEADER.DATA_DIRECTORY[pefile.DIRECTORY_ENTRY['IMAGE_DIRECTORY_ENTRY_SECURITY']]
        if sec_dir.VirtualAddress != 0:
            sig_start = sec_dir.VirtualAddress
            print(f"[!] Digital signature found at byte {sig_start}, size {sec_dir.Size}")
            pe_end_offset = max(pe_end_offset, sig_start + sec_dir.Size)

        with open(exe_path, "rb") as f:
            f.seek(0, 2)
            total_size = f.tell()

            if total_size <= pe_end_offset:
                print("[-] No overlay found")
                pe.close()
                return

            overlay_size = total_size - pe_end_offset
            print(f"[+] Overlay size: {overlay_size} bytes")

            f.seek(pe_end_offset)
            overlay_data = f.read()

        with open(output_path, "wb") as out:
            out.write(overlay_data)
        print(f"[+] Saved to: {output_path}")

        pe.close()

    except Exception as e:
        print(f"[-] Error: {e}")

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python overlay_extract.py <file.exe> <output.bin>")
        sys.exit(1)
    extract_overlay(sys.argv[1], sys.argv[2])